# Classification

### - Analyze data Bankloan.csv
        - Target: Default
        - Features: Employ, Debtinc, Creddebt, othdebt
### - Interpret The Result
### - Validate the model using accuracy in 20% testing data

## Data

In [1]:
import pandas as pd
import numpy as np

In [2]:
bankloan = pd.read_csv(r'C:\Users\dheof\Desktop\Help\Purwadhika\Csv_Files\bankloan.csv')
bankloan

age  ed  employ  address  income  debtinc   creddebt   othdebt  default
0     41   3      17       12     176      9.3  11.359392  5.008608        1
1     27   1      10        6      31     17.3   1.362202  4.000798        0
2     40   1      15       14      55      5.5   0.856075  2.168925        0
3     41   1      15       14     120      2.9   2.658720  0.821280        0
4     24   2       2        0      28     17.3   1.787436  3.056564        1
..   ...  ..     ...      ...     ...      ...        ...       ...      ...
695   36   2       6       15      27      4.6   0.262062  0.979938        1
696   29   2       6        4      21     11.5   0.369495  2.045505        0
697   33   1      15        3      32      7.6   0.491264  1.940736        0
698   45   1      19       22      77      8.4   2.302608  4.165392        0
699   37   1      12       14      44     14.7   2.994684  3.473316        0

[700 rows x 9 columns]

### Keterangan
        - Age       : Usia
        - Ed        : Jenjang pendidikan (1,2,3...)
        - Employ    : Lama Kerja
        - Address   : Lama tinggal
        - Income    : Pemasukan per bulan
        - Debtinc   : Persentase utang dalam income
        - Creddebt  : Nominal utang (1000 dollar)
        - Othdebt   : Nominal utang dari sumber lain (1000 dollar)


In [3]:
fitur = ['employ', 'debtinc', 'creddebt', 'othdebt']
target = 'default'

x = bankloan[fitur]
y = bankloan[target]

In [4]:
x.describe()

employ     debtinc    creddebt     othdebt
count  700.000000  700.000000  700.000000  700.000000
mean     8.388571   10.260571    1.553553    3.058209
std      6.658039    6.827234    2.117197    3.287555
min      0.000000    0.400000    0.011696    0.045584
25%      3.000000    5.000000    0.369059    1.044178
50%      7.000000    8.600000    0.854869    1.987567
75%     12.000000   14.125000    1.901955    3.923065
max     31.000000   41.300000   20.561310   27.033600

## Logistic Regression

In [5]:
import statsmodels.api as sm

In [6]:
model = sm.Logit(y, sm.add_constant(x)) #Definisi Model
result = model.fit()

Optimization terminated successfully.
         Current function value: 0.411165
         Iterations 7


In [7]:
print(result.summary())

                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                  700
Model:                          Logit   Df Residuals:                      695
Method:                           MLE   Df Model:                            4
Date:                Tue, 21 Sep 2021   Pseudo R-squ.:                  0.2844
Time:                        16:08:58   Log-Likelihood:                -287.82
converged:                       True   LL-Null:                       -402.18
Covariance Type:            nonrobust   LLR p-value:                 2.473e-48
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.2302      0.236     -5.210      0.000      -1.693      -0.767
employ        -0.2436      0.029     -8.456      0.000      -0.300      -0.187
debtinc        0.0885      0.021      4.200      0.0

### Output:
### 1. LLR P-Value: 0.0000....2473
### 2. P-Value:
        - Const     : 0.000
        - Employ    : 0.000
        - Debtinc   : 0.000
        - Creddebt  : 0.000
        - Othdebt   : 0.940
### 3. Coef: 
        - Employ    : -0.2436
        - Debtinc   : 0.0885
        - Creddebt  : 0.5041
        - Othdebt   : -0.0040

In [8]:
bankloan

age  ed  employ  address  income  debtinc   creddebt   othdebt  default
0     41   3      17       12     176      9.3  11.359392  5.008608        1
1     27   1      10        6      31     17.3   1.362202  4.000798        0
2     40   1      15       14      55      5.5   0.856075  2.168925        0
3     41   1      15       14     120      2.9   2.658720  0.821280        0
4     24   2       2        0      28     17.3   1.787436  3.056564        1
..   ...  ..     ...      ...     ...      ...        ...       ...      ...
695   36   2       6       15      27      4.6   0.262062  0.979938        1
696   29   2       6        4      21     11.5   0.369495  2.045505        0
697   33   1      15        3      32      7.6   0.491264  1.940736        0
698   45   1      19       22      77      8.4   2.302608  4.165392        0
699   37   1      12       14      44     14.7   2.994684  3.473316        0

[700 rows x 9 columns]

### Yang dimodelkan adalah risiko default, risiko seseorang akan menjadi buruk pembayarannya.
### Interpretasi:
### 1. LLR P-Value: 0.0000....2473 < 0.05 (Tolak H0) Artinya minimal salah satu variable berpengaruh secara signifikan terhadap risiko default
### 2. P-Value:
        - Const     : 0.000 < 0.05 (Tolak H0). Artinya mode memerlukan intersep
        - Employ    : 0.000 < 0.05 (Tolak H0) Artinya, lama kerja berpengaruh secara negatif terhadap risiko default. Semakin lama kerja semakin rendah risiko default seseorang. Berlaku untuk Emplot dari (0 - 31)
        - Debtinc   : 0.000 < 0.05 (Tolak H0) Artinya persentae hutang berpengaruh secara positif terhadap risiko default. Semakin besar persentase utang semakin tingg juga risiko default. 
        Berlaku untuk Debtinc antara 0.4 - 41.3 (1000 dollar)
        - Creddebt  : 0.000 < 0.05 (Tolak H0) Artinya nominal hutang berpengaruh secara positif terhadap risiko default. Semakin besar nominal hutang semakin tinggi juga risiko default. 
        Berlaku untuk Creddebt antara 0.011696 - 20.561310 (1000 dollar)
        - Othdebt   : 0.940 > 0.05 (Terima H0) Artinya tidak cukup bukti bahwa nominal utang dari sumber lain berpengaruh terhadap risiko default. Berlaku untuk Othdebt antara 0.045584 - 27.033600 (1000 dollar)
### 3. Coef: 
        - Employ    : -0.2436. Ketika lama kerja meningnkat 5 tahun, kecenderungan sesorang akan menjadi default menurun sebesar 3.38 kali lipat.(Nilai variabel lain tidak berubah)        
        - Debtinc   : 0.0885. Ketika persentase utang meningkat sebesar 5%, kecenderungan seseorang akan menjadu default meningkat sebesar 1.556 kali lipat. (Nilai variabel lain tidak berubah)
        - Creddebt  : 0.5041. Ketika nominal utang meningkat sebesar 5000 dollar, kecenderungan seseorang akan menjadi default meningkat sebesar 12.434 kali lipat. (Nilai variabel lain tidak berubah)
        - Othdebt   : -0.0040. Ketika nominal hutang dari sumber lain meningkat 5000 dollar, kecenderungan seseorang akan menjadi default menurun sebesar 1.0202 kali lipat. (Nilai variabel lain tidak berubah)      

### Coef --> OR exp(beta * (c - a))

In [9]:
## EMPLOY
c = 20 
a = 15

OR_employ = np.exp(0.2436 * (c - a))
print(OR_employ)

3.380420128015566


In [10]:
## DEBTINC
c = 20 
a = 15

OR_debtinc = np.exp(0.0885 * (c - a))
print(OR_debtinc)

1.5565938428137092


In [11]:
## CREDDEBT
c = 20 
a = 15

OR_creddebt = np.exp(0.5041 * (c - a))
print(OR_creddebt)

12.434812515742879


In [12]:
## OTHDEBT
c = 20 
a = 15

OR_othdebt = np.exp(0.0040 * (c - a))
print(OR_othdebt)

1.0202013400267558


## Multikol

In [13]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [14]:
def calc_vif(x):
    vif = pd.DataFrame()
    vif['variabel'] = x.columns
    vif['vif'] = [variance_inflation_factor(x.values, i) for i in range(x.shape[1])]
    
    return vif

In [15]:
calc_vif(x)

variabel       vif
0    employ  2.222753
1   debtinc  3.045977
2  creddebt  2.816577
3   othdebt  4.116876

## Validasi Model

### Validasi pada 20% data testing

In [16]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [17]:
x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size = 0.2,
    random_state = 2020
)

In [18]:
sm_Logit_train = sm.Logit(y_train, sm.add_constant(x_train))
result_train = sm_Logit_train.fit()

Optimization terminated successfully.
         Current function value: 0.408607
         Iterations 7


In [19]:
y_prob = result_train.predict(sm.add_constant(x_test)) # RISIKO/PELUANG DEFAULT
y_class = np.where(y_prob > 0.5, 1, 0) ## KELAS DEFAULT ATAU NON DEFAULT

In [20]:
print('Akurasinya adalah sebesar', accuracy_score(y_test, y_class) * 100, '%')


Akurasinya adalah sebesar 82.14285714285714 %


### Jika ada 100 nasabah, sekitar 82 nasabah akan terprediksi (default/non-default) dengan benar